### ISSUE TITLE
Sensu Not Passing Data to Elasticsearch Handler

### ISSUE THREAD

**Issue by davpx** [sensu/sensu-community-plugins#1177]
Hello,

Hoping someone can help. I'm unable to get the elasticsearch handler working. I've seen many implementations where some people use a pipe handler and some just have an elasticsearch.json with the elasticsearch cluster ip, port, index, and timeout. If I place the elasticsearch_metrics.rb file in /etc/sensu/extensions/handlers , I can see sensu-server pick up the file but the service crashes on restart and kills my shell. If I leave it in /etc/sensu/handlers/metrics/elasticsearch_metrics.rb, sensu-server starts just fine.

Regardless of my config, I never get data into elasticsearch. Trying to send checks and metrics to see if anything comes through but no avail.

handler config:
/etc/sensu/conf.d/handlers/elasticsearch_metrics.json
{
  "elasticsearch": {
    "host": "localhost",
    "port": 9200,
    "es_index": "sensu-metrics",
    "es_type": "metrics",
    "timeout": 5
  }
}

also tried adding a pipe handler but no good
/etc/sensu/conf.d/handlers/elasticsearch.json
{
 "handlers": {
  "elasticsearch": {
   "type": "pipe",
   "command": "/etc/sensu/handlers/metrics/elasticsearch_metrics.rb"
   "timeout": 5
  }
 }
}

I do see the following events in sensu-server.log but I can confirm that the elasticsearch port is open and accepting communications.

{"timestamp":"2015-05-12T18:55:24.886508-0400","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/sensu/handlers/metrics/elasticsearch_metrics.rb","timeout":5,"host":"localhost","port":9200,"index":"sensu-metrics","name":"elasticsearch"},"output":["/opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `initialize': Connection refused - connect(2) (Errno::ECONNREFUSED)\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `open'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `block in connect'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:52:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:877:in `connect'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:862:in `do_start'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:851:in `start'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:1367:in `request'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:85:in `api_request'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:120:in `stash_exists?'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:132:in `block (2 levels) in filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:66:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:97:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:131:in `block in filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:129:in `each'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:129:in `filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:31:in `filter'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:54:in `block in <class:Handler>'\n"]}

**Comment by portertech** [comment]
The community ES handler that you are referring to is not an extension, therefore, do not put it in `/etc/sensu/extensions`. It looks like you are missing the API configuration snippet the `sensu-plugin` requires to communicate with the Sensu API, e.g. `{"api": {"host": "localhost", "port": 4567}}`

**Comment by davpx** [comment]
I've added the api config to /etc/sensu/handlers/metrics/elasticsearch_metrics.json after also experimenting with this file in /etc/sensu/conf.d/handlers/ but no luck.
{
  "elasticsearch": {
    "host": "localhost",
    "port": 9200,
    "index": "sensu-metrics",
    "type": "metrics",
    "timeout": 5
  },
  "api": {
   "host": "localhost",
   "port": 4567
 }
}

I still have /etc/sensu/conf.d/handlers/elasticsearch.json present which is still throwing CONN REFUSED errors
{
 "handlers": {
  "elasticsearch": {
   "type": "pipe",
   "command": "/etc/sensu/handlers/metrics/elasticsearch_metrics.rb",
   "timeout": 5
  }
 }
}
{"timestamp":"2015-05-12T19:43:14.783240-0400","level":"info","message":"handler output","handler":{"type":"pipe","command":"/etc/sensu/handlers/metrics/elasticsearch_metrics.rb","timeout":5,"name":"elasticsearch"},"output":["/opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `initialize': Connection refused - connect(2) (Errno::ECONNREFUSED)\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `open'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `block in connect'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:52:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:877:in `connect'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:862:in `do_start'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:851:in `start'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:1367:in `request'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:85:in `api_request'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:120:in `stash_exists?'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:132:in `block (2 levels) in filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:66:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:97:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:131:in `block in filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:129:in `each'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:129:in `filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:31:in `filter'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:54:in `block in <class:Handler>'\n"]}

**Comment by portertech** [comment]
Is the Sensu API running on the same machine as this Sensu server? Does it use authentication?

**Comment by davpx** [comment]
Yes is is running one the same server and no authentication currently on the api. I turned that off awhile back to rule it out.

**Comment by pdaugavietis** [comment]
you shouldn't need the api config there, if it's in another config file anyway -- but make sure that elasticsearch is listening on the right interface(s) and there aren't any funny firewalls in between...   connection refused sounds like something either isn't listening to where you expect it to be, or that something is actively killing the connection before is has a chance.

**Comment by davpx** [comment]
Here is a summary of my current state. connection refused errors are back. Confirmed no firewall in place, elasticsearch listening on port 9200 and I can pipe data directly into it from my shell using curl.

Logs Showing Handler Config Accepted (/var/log/sensu/sensu-server.log):
```
{"timestamp":"2015-05-15T13:11:19.393381-0400","level":"warn","message":"config file applied changes","file":"/etc/sensu/conf.d/handlers/elasticsearch.json","changes":{"handlers":{"elasticsearch":[null,{"type":"pipe","command":"/opt/sensu/embedded/bin/ruby /etc/sensu/handlers/metrics/elasticsearch_metrics.rb","host":"localhost","port":9200,"index":"sensu-metrics","timeout":5}]}}}

{"timestamp":"2015-05-15T13:11:19.393475-0400","level":"warn","message":"loading config file","file":"/etc/sensu/conf.d/handlers/handler_notification.json"}

{"timestamp":"2015-05-15T13:11:19.393634-0400","level":"warn","message":"config file applied changes","file":"/etc/sensu/conf.d/handlers/handler_notification.json","changes":{"handlers":{"sns":[null,{"type":"pipe","command":"/etc/sensu/handlers/sns.rb"}],"mailer":[null,{"type":"pipe","command":"/etc/sensu/handlers/mailer.rb"}],"pagerduty":[null,{"type":"pipe","command":"/etc/sensu/handlers/notification/pagerduty.rb"}]}}}

{"timestamp":"2015-05-15T13:11:19.393758-0400","level":"warn","message":"loading config file","file":"/etc/sensu/conf.d/handlers/pagerduty.json"}

{"timestamp":"2015-05-15T13:11:19.393877-0400","level":"warn","message":"config file applied changes","file":"/etc/sensu/conf.d/handlers/pagerduty.json","changes":{"pagerduty":[null,{"api_key":"REDACTED","Splunk":{"api_key":"REDACTED"},"team_name2":{"api_key":"REDACTED"}}]}}
```

Logs Showing Handler Error (/var/log/sensu/sensu-server.log):
```
{"timestamp":"2015-05-15T13:25:04.875344-0400","level":"info","message":"handler output","handler":{"type":"pipe","command":"/opt/sensu/embedded/bin/ruby /etc/sensu/handlers/metrics/elasticsearch_metrics.rb","host":"localhost","port":9200,"index":"sensu-metrics","timeout":5,"name":"elasticsearch"},"output":["/opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `initialize': Connection refused - connect(2) (Errno::ECONNREFUSED)\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `open'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:878:in `block in connect'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:52:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:877:in `connect'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:862:in `do_start'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:851:in `start'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/net/http.rb:1373:in `request'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:85:in `api_request'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:120:in `stash_exists?'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:132:in `block (2 levels) in filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:66:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/2.0.0/timeout.rb:97:in `timeout'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:131:in `block in filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:129:in `each'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:129:in `filter_silenced'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:31:in `filter'\n","\tfrom /opt/sensu/embedded/lib/ruby/gems/2.0.0/gems/sensu-plugin-1.1.0/lib/sensu-handler.rb:54:in `block in <class:Handler>'\n"]}
```

Telnet Results to 9200:
```
telnet localhost 9200
Trying ::1...
Connected to localhost.
Escape character is '^]'.
```

Elasticsearch Handler Config (/etc/sensu/conf.d/handlers/elasticsearch.json):
```
{
 "handlers": {
  "elasticsearch": {
   "type": "pipe",
   "command": "/opt/sensu/embedded/bin/ruby /etc/sensu/handlers/metrics/elasticsearch_metrics.rb",
   "host": "localhost",
   "port": 9200,
   "index": "sensu-metrics",
   "timeout": 5
  }
 }
}
```

API Config (/etc/sensu/conf.d/api.json):
```
{
  "api": {
    "host": "localhost",
    "port": 4567
  }
}
```

RabbitMQ Config (/etc/sensu/conf.d/rabbitmq.json):
```
{
  "rabbitmq": {
    "ssl": {
      "cert_chain_file": "/etc/sensu/ssl/cert.pem",
      "private_key_file": "/etc/sensu/ssl/key.pem"
    },
    "host": "localhost",
    "port": 5671,
    "vhost": "/sensu",
    "user": "sensu",
    "password": "<REDACTED>"
  }
}
```

Redis Config (/etc/sensu/conf.d/redis.json):
```
{
  "redis": {
    "host": "localhost",
    "port": 6379
  }
}
```

Global Check Config Snippet (/etc/sensu/conf.d/checks/global.json):
```
{
  "checks": {
    "cpu_check": {
      "handlers": ["default", "elasticsearch"],
      "command": "/opt/sensu/sensu-community-plugins/plugins/system/check-cpu.rb -w 80 -c 90 ",
      "interval": 10,
      "subscribers": [ "global" ]
    },
    "cpu_metrics": {
     "handlers": ["elasticsearch"],
     "command": "/opt/sensu/sensu-community-plugins/plugins/system/cpu-metrics.rb",
     "interval": 10,
     "subscribers": [ "global" ]
    }
  }
}
```

**Comment by davpx** [comment]
One more update. So I tried passing a sensu check output manually to the elasticsearch handler. I opened a new nc listening port and piped a few check outputs to it using nc as well. The check output I got is as below. Notice that there is no "elasticsearch" block in the check output. Why is sensu-server leaving this out with the config as above?

`
{"id":"59b9ccf9-7025-469c-9496-0387ad5c1382","client":{"name":"sensu-test","address":"localhost","subscriptions":["global"],"version":"0.18.1","timestamp":1431712810},"check":{"handlers":["elasticsearch"],"command":"/opt/sensu/embedded/bin/ruby /opt/sensu-community-plugins/plugins/system/cpu-metrics.rb","interval":10,"runbook":"http://woki/x/rkk7Eg","team":"MSI","project":"SAAS","type":"metric","subscribers":["global"],"name":"cpu_metrics","issued":1431712818,"executed":1431712818,"duration":0.201,"output":"sensu-test.cpu.total.user 4850 1431712818\nsensu-test.cpu.total.nice 0 1431712818\nsensu-test.cpu.total.system 1256 1431712818\nsensu-test.cpu.total.idle 511802 1431712818\nsensu-test.cpu.total.iowait 537 1431712818\nsensu-test.cpu.total.irq 169 1431712818\nsensu-test.cpu.total.softirq 87 1431712818\nsensu-test.cpu.total.steal 0 1431712818\nsensu-test.cpu.total.guest 0 1431712818\nsensu-test.cpu.cpu0.user 1489 1431712818\nsensu-test.cpu.cpu0.nice 0 1431712818\nsensu-test.cpu.cpu0.system 437 1431712818\nsensu-test.cpu.cpu0.idle 126282 1431712818\nsensu-test.cpu.cpu0.iowait 280 1431712818\nsensu-test.cpu.cpu0.irq 169 1431712818\nsensu-test.cpu.cpu0.softirq 57 1431712818\nsensu-test.cpu.cpu0.steal 0 1431712818\nsensu-test.cpu.cpu0.guest 0 1431712818\nsensu-test.cpu.cpu1.user 1034 1431712818\nsensu-test.cpu.cpu1.nice 0 1431712818\nsensu-test.cpu.cpu1.system 289 1431712818\nsensu-test.cpu.cpu1.idle 128035 1431712818\nsensu-test.cpu.cpu1.iowait 76 1431712818\nsensu-test.cpu.cpu1.irq 0 1431712818\nsensu-test.cpu.cpu1.softirq 16 1431712818\nsensu-test.cpu.cpu1.steal 0 1431712818\nsensu-test.cpu.cpu1.guest 0 1431712818\nsensu-test.cpu.cpu2.user 854 1431712818\nsensu-test.cpu.cpu2.nice 0 1431712818\nsensu-test.cpu.cpu2.system 263 1431712818\nsensu-test.cpu.cpu2.idle 128817 1431712818\nsensu-test.cpu.cpu2.iowait 67 1431712818\nsensu-test.cpu.cpu2.irq 0 1431712818\nsensu-test.cpu.cpu2.softirq 8 1431712818\nsensu-test.cpu.cpu2.steal 0 1431712818\nsensu-test.cpu.cpu2.guest 0 1431712818\nsensu-test.cpu.cpu3.user 1472 1431712818\nsensu-test.cpu.cpu3.nice 0 1431712818\nsensu-test.cpu.cpu3.system 265 1431712818\nsensu-test.cpu.cpu3.idle 128666 1431712818\nsensu-test.cpu.cpu3.iowait 112 1431712818\nsensu-test.cpu.cpu3.irq 0 1431712818\nsensu-test.cpu.cpu3.softirq 4 1431712818\nsensu-test.cpu.cpu3.steal 0 1431712818\nsensu-test.cpu.cpu3.guest 0 1431712818\nsensu-test.cpu.intr 0 1431712818\nsensu-test.cpu.ctxt 430349 1431712818\nsensu-test.cpu.btime 1431711507 1431712818\nsensu-test.cpu.processes 3568 1431712818\nsensu-test.cpu.procs_running 1 1431712818\nsensu-test.cpu.procs_blocked 0 1431712818\nsensu-test.cpu.cpu_count 4 1431712818\n\n","status":0,"history":["0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0","0"],"total_state_change":0},"occurrences":1}
`
